<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Overview of the tidyseurat package • tidyseurat</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Overview of the tidyseurat package">
<meta property="og:description" content="tidyseurat">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">tidyseurat</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.11</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/introduction.html">Overview of the tidyseurat package</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Overview of the tidyseurat package</h1>
                        <h4 class="author">Stefano Mangiola</h4>
            
            <h4 class="date">2020-10-20</h4>
      
      
      <div class="hidden name"><code>introduction.Rmd</code></div>

    </div>

    
    
<!-- badges: start -->
<p><a href="https://www.tidyverse.org/lifecycle/#maturing"><img src="https://img.shields.io/badge/lifecycle-maturing-blue.svg" alt="Lifecycle:maturing"></a> <!-- badges: end --></p>
<p><strong>Brings Seurat to the tidyverse!</strong></p>
<p>website: <a href="https://stemangiola.github.io/tidyseurat/">stemangiola.github.io/tidyseurat/</a></p>
<p>Please also have a look at</p>
<ul>
<li>
<a href="https://stemangiola.github.io/tidybulk/">tidybulk</a> for tidy bulk RNA-seq analysis</li>
<li>
<a href="https://github.com/stemangiola/nanny">nanny</a> for tidy high-level data analysis and manipulation</li>
<li>
<a href="https://github.com/stemangiola/tidygate">tidygate</a> for adding custom gate information to your tibble</li>
<li>
<a href="https://stemangiola.github.io/tidyHeatmap/">tidyHeatmap</a> for heatmaps produced with tidy principles</li>
</ul>
<div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
<p>tidyseurat provides a bridge between the Seurat single-cell package <span class="citation">(Butler et al. 2018; Stuart et al. 2019)</span> and the tidyverse <span class="citation">(Wickham et al. 2019)</span>. It creates an invisible layer that enables viewing the Seurat object as a tidyverse tibble, and provides Seurat-compatible <em>dplyr</em>, <em>tidyr</em>, <em>ggplot</em> and <em>plotly</em> functions.</p>
<div id="functionsutilities-available" class="section level2">
<h2 class="hasAnchor">
<a href="#functionsutilities-available" class="anchor"></a>Functions/utilities available</h2>
<table class="table">
<thead><tr class="header">
<th>Seurat-compatible Functions</th>
<th>Description</th>
</tr></thead>
<tbody><tr class="odd">
<td><code>all</code></td>
<td>After all <code>tidyseurat</code> is a Seurat object, just better</td>
</tr></tbody>
</table>
<table class="table">
<thead><tr class="header">
<th>tidyverse Packages</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><code>dplyr</code></td>
<td>All <code>dplyr</code> APIs like for any tibble</td>
</tr>
<tr class="even">
<td><code>tidyr</code></td>
<td>All <code>tidyr</code> APIs like for any tibble</td>
</tr>
<tr class="odd">
<td><code>ggplot2</code></td>
<td>
<code>ggplot</code> like for any tibble</td>
</tr>
<tr class="even">
<td><code>plotly</code></td>
<td>
<code>plot_ly</code> like for any tibble</td>
</tr>
</tbody>
</table>
<table class="table">
<thead><tr class="header">
<th>Utilities</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><code>tidy</code></td>
<td>Add <code>tidyseurat</code> invisible layer over a Seurat object</td>
</tr>
<tr class="even">
<td><code>as_tibble</code></td>
<td>Convert cell-wise information to a <code>tbl_df</code>
</td>
</tr>
<tr class="odd">
<td><code>join_transcripts</code></td>
<td>Add transcript-wise information, returns a <code>tbl_df</code>
</td>
</tr>
</tbody>
</table>
</div>
<div id="installation" class="section level2">
<h2 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h2>
<p>From CRAN</p>
<div class="sourceCode" id="cb1"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span>(<span class="st">"tidyseurat"</span>)</pre></body></html></div>
<p>From Github (development)</p>
<div class="sourceCode" id="cb2"><html><body><pre class="r"><span class="kw pkg">devtools</span><span class="kw ns">::</span><span class="fu"><a href="https://devtools.r-lib.org//reference/remote-reexports.html">install_github</a></span>(<span class="st">"stemangiola/tidyseurat"</span>)</pre></body></html></div>
<div class="sourceCode" id="cb3"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">dplyr</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">tidyr</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">purrr</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">magrittr</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">ggplot2</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">Seurat</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">tidyseurat</span>)</pre></body></html></div>
</div>
<div id="create-tidyseurat-the-best-of-both-worlds" class="section level2">
<h2 class="hasAnchor">
<a href="#create-tidyseurat-the-best-of-both-worlds" class="anchor"></a>Create <code>tidyseurat</code>, the best of both worlds!</h2>
<p>This is a seurat object but it is evaluated as tibble. So it is fully compatible both with Seurat and tidyverse APIs.</p>
<div class="sourceCode" id="cb4"><html><body><pre class="r"><span class="no">pbmc_small_tidy</span> <span class="kw">&lt;-</span> <span class="kw pkg">tidyseurat</span><span class="kw ns">::</span><span class="no"><a href="../reference/pbmc_small.html">pbmc_small</a></span> <span class="kw">%&gt;%</span> <span class="fu"><a href="../reference/tidy.html">tidy</a></span>()</pre></body></html></div>
<p><strong>It looks like a tibble</strong></p>
<div class="sourceCode" id="cb5"><html><body><pre class="r"><span class="no">pbmc_small_tidy</span></pre></body></html></div>
<pre><code>## # A tibble abstraction: 80 x 16
##    cell  orig.ident nCount_RNA nFeature_RNA RNA_snn_res.0.8 letter.idents groups
##    &lt;chr&gt; &lt;fct&gt;           &lt;dbl&gt;        &lt;int&gt; &lt;fct&gt;           &lt;fct&gt;         &lt;chr&gt; 
##  1 ATGC… SeuratPro…         70           47 0               A             g2    
##  2 CATG… SeuratPro…         85           52 0               A             g1    
##  3 GAAC… SeuratPro…         87           50 1               B             g2    
##  4 TGAC… SeuratPro…        127           56 0               A             g2    
##  5 AGTC… SeuratPro…        173           53 0               A             g2    
##  6 TCTG… SeuratPro…         70           48 0               A             g1    
##  7 TGGT… SeuratPro…         64           36 0               A             g1    
##  8 GCAG… SeuratPro…         72           45 0               A             g1    
##  9 GATA… SeuratPro…         52           36 0               A             g1    
## 10 AATG… SeuratPro…        100           41 0               A             g1    
## # … with 70 more rows, and 9 more variables: RNA_snn_res.1 &lt;fct&gt;, file &lt;chr&gt;,
## #   PC_1 &lt;dbl&gt;, PC_2 &lt;dbl&gt;, PC_3 &lt;dbl&gt;, PC_4 &lt;dbl&gt;, PC_5 &lt;dbl&gt;, tSNE_1 &lt;dbl&gt;,
## #   tSNE_2 &lt;dbl&gt;</code></pre>
<p><strong>But it is a Seurat object after all</strong></p>
<div class="sourceCode" id="cb7"><html><body><pre class="r"><span class="no">pbmc_small_tidy</span>@<span class="kw">assays</span></pre></body></html></div>
<pre><code>## $RNA
## Assay data with 230 features for 80 cells
## Top 10 variable features:
##  PPBP, IGLL5, VDAC3, CD1C, AKR1C3, PF4, MYL9, GNLY, TREML1, CA2</code></pre>
</div>
</div>
<div id="annotation-polishing" class="section level1">
<h1 class="hasAnchor">
<a href="#annotation-polishing" class="anchor"></a>Annotation polishing</h1>
<p>We may have a column that contains the directory each run was taken from, such as the “file” column in <code>pbmc_small_tidy</code>.</p>
<div class="sourceCode" id="cb9"><html><body><pre class="r"><span class="no">pbmc_small_tidy</span>$<span class="no">file</span>[<span class="fl">1</span>:<span class="fl">5</span>]</pre></body></html></div>
<pre><code>##                                     ATGCCAGAACGACT 
## "../data/sample2/outs/filtered_feature_bc_matrix/" 
##                                     CATGGCCTGTGCAT 
## "../data/sample1/outs/filtered_feature_bc_matrix/" 
##                                     GAACCTGATGAACC 
## "../data/sample2/outs/filtered_feature_bc_matrix/" 
##                                     TGACTGGATTCTCA 
## "../data/sample2/outs/filtered_feature_bc_matrix/" 
##                                     AGTCAGACTGCACA 
## "../data/sample2/outs/filtered_feature_bc_matrix/"</code></pre>
<p>We may want to extract the run/sample name out of it into a separate column. Tidyverse <code>extract</code> can be used to convert a character column into multiple columns using regular expression groups.</p>
<div class="sourceCode" id="cb11"><html><body><pre class="r"><span class="co"># Create sample column</span>
<span class="no">pbmc_small_polished</span> <span class="kw">&lt;-</span>
  <span class="no">pbmc_small_tidy</span> <span class="kw">%&gt;%</span>
  <span class="fu"><a href="../reference/tidyr-methods.html">extract</a></span>(<span class="no">file</span>, <span class="st">"sample"</span>, <span class="st">"../data/([a-z0-9]+)/outs.+"</span>, <span class="kw">remove</span> <span class="kw">=</span> <span class="fl">FALSE</span>)
<span class="co"># Reorder to have sample column up front</span>
<span class="no">pbmc_small_polished</span> <span class="kw">%&gt;%</span>
  <span class="fu"><a href="../reference/dplyr-methods.html">select</a></span>(<span class="no">sample</span>, <span class="fu"><a href="https://dplyr.tidyverse.org/reference/reexports.html">everything</a></span>())</pre></body></html></div>
<pre><code>## # A tibble abstraction: 80 x 17
##    cell  sample orig.ident nCount_RNA nFeature_RNA RNA_snn_res.0.8 letter.idents
##    &lt;chr&gt; &lt;chr&gt;  &lt;fct&gt;           &lt;dbl&gt;        &lt;int&gt; &lt;fct&gt;           &lt;fct&gt;        
##  1 ATGC… sampl… SeuratPro…         70           47 0               A            
##  2 CATG… sampl… SeuratPro…         85           52 0               A            
##  3 GAAC… sampl… SeuratPro…         87           50 1               B            
##  4 TGAC… sampl… SeuratPro…        127           56 0               A            
##  5 AGTC… sampl… SeuratPro…        173           53 0               A            
##  6 TCTG… sampl… SeuratPro…         70           48 0               A            
##  7 TGGT… sampl… SeuratPro…         64           36 0               A            
##  8 GCAG… sampl… SeuratPro…         72           45 0               A            
##  9 GATA… sampl… SeuratPro…         52           36 0               A            
## 10 AATG… sampl… SeuratPro…        100           41 0               A            
## # … with 70 more rows, and 10 more variables: groups &lt;chr&gt;,
## #   RNA_snn_res.1 &lt;fct&gt;, file &lt;chr&gt;, PC_1 &lt;dbl&gt;, PC_2 &lt;dbl&gt;, PC_3 &lt;dbl&gt;,
## #   PC_4 &lt;dbl&gt;, PC_5 &lt;dbl&gt;, tSNE_1 &lt;dbl&gt;, tSNE_2 &lt;dbl&gt;</code></pre>
</div>
<div id="preliminary-plots" class="section level1">
<h1 class="hasAnchor">
<a href="#preliminary-plots" class="anchor"></a>Preliminary plots</h1>
<p>Set colours and theme for plots.</p>
<div class="sourceCode" id="cb13"><html><body><pre class="r"><span class="co"># Use colourblind-friendly colours</span>
<span class="kw">if</span> (<span class="fu"><a href="https://rdrr.io/r/base/ns-load.html">requireNamespace</a></span>(<span class="st">"dittoSeq"</span>, <span class="kw">quietly</span> <span class="kw">=</span> <span class="fl">TRUE</span>)) {
      <span class="no">friendly_cols</span> <span class="kw">&lt;-</span> <span class="kw pkg">dittoSeq</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/dittoSeq/man/dittoColors.html">dittoColors</a></span>()
   } <span class="kw">else</span> {
      <span class="no">friendly_cols</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"red"</span>, <span class="st">"blue"</span>, <span class="st">"green"</span>, <span class="st">"purple"</span>)
   }

<span class="co"># Set theme</span>
<span class="no">my_theme</span> <span class="kw">&lt;-</span>
  <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span>(
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_fill_manual</a></span>(<span class="kw">values</span> <span class="kw">=</span> <span class="no">friendly_cols</span>),
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_color_manual</a></span>(<span class="kw">values</span> <span class="kw">=</span> <span class="no">friendly_cols</span>),
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span>() +
      <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span>(
        <span class="kw">panel.border</span> <span class="kw">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_blank</a></span>(),
        <span class="kw">axis.line</span> <span class="kw">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_line</a></span>(),
        <span class="kw">panel.grid.major</span> <span class="kw">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_line</a></span>(<span class="kw">size</span> <span class="kw">=</span> <span class="fl">0.2</span>),
        <span class="kw">panel.grid.minor</span> <span class="kw">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_line</a></span>(<span class="kw">size</span> <span class="kw">=</span> <span class="fl">0.1</span>),
        <span class="kw">text</span> <span class="kw">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span>(<span class="kw">size</span> <span class="kw">=</span> <span class="fl">12</span>),
        <span class="kw">legend.position</span> <span class="kw">=</span> <span class="st">"bottom"</span>,
        <span class="kw">aspect.ratio</span> <span class="kw">=</span> <span class="fl">1</span>,
        <span class="kw">strip.background</span> <span class="kw">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_blank</a></span>(),
        <span class="kw">axis.title.x</span> <span class="kw">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span>(<span class="kw">margin</span> <span class="kw">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">margin</a></span>(<span class="kw">t</span> <span class="kw">=</span> <span class="fl">10</span>, <span class="kw">r</span> <span class="kw">=</span> <span class="fl">10</span>, <span class="kw">b</span> <span class="kw">=</span> <span class="fl">10</span>, <span class="kw">l</span> <span class="kw">=</span> <span class="fl">10</span>)),
        <span class="kw">axis.title.y</span> <span class="kw">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span>(<span class="kw">margin</span> <span class="kw">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">margin</a></span>(<span class="kw">t</span> <span class="kw">=</span> <span class="fl">10</span>, <span class="kw">r</span> <span class="kw">=</span> <span class="fl">10</span>, <span class="kw">b</span> <span class="kw">=</span> <span class="fl">10</span>, <span class="kw">l</span> <span class="kw">=</span> <span class="fl">10</span>))
      )
  )</pre></body></html></div>
<p>We can treat <code>pbmc_small_polished</code> effectively as a normal tibble for plotting.</p>
<p>Here we plot number of transcripts per cell.</p>
<div class="sourceCode" id="cb14"><html><body><pre class="r"><span class="no">pbmc_small_polished</span> <span class="kw">%&gt;%</span>
  <span class="kw pkg">tidyseurat</span><span class="kw ns">::</span><span class="fu"><a href="../reference/ggplot2-methods.html">ggplot</a></span>(<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="no">nFeature_RNA</span>, <span class="kw">fill</span> <span class="kw">=</span> <span class="no">groups</span>)) +
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html">geom_histogram</a></span>() +
  <span class="no">my_theme</span></pre></body></html></div>
<p><img src="introduction_files/figure-html/plot1-1.png" width="700"></p>
<p>Here we plot total transcripts per cell.</p>
<div class="sourceCode" id="cb15"><html><body><pre class="r"><span class="no">pbmc_small_polished</span> <span class="kw">%&gt;%</span>
  <span class="kw pkg">tidyseurat</span><span class="kw ns">::</span><span class="fu"><a href="../reference/ggplot2-methods.html">ggplot</a></span>(<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="no">groups</span>, <span class="no">nCount_RNA</span>, <span class="kw">fill</span> <span class="kw">=</span> <span class="no">groups</span>)) +
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html">geom_boxplot</a></span>(<span class="kw">outlier.shape</span> <span class="kw">=</span> <span class="fl">NA</span>) +
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_jitter.html">geom_jitter</a></span>(<span class="kw">width</span> <span class="kw">=</span> <span class="fl">0.1</span>) +
  <span class="no">my_theme</span></pre></body></html></div>
<p><img src="introduction_files/figure-html/plot2-1.png" width="700"></p>
<p>Here we plot abundance of two transcripts for each group.</p>
<div class="sourceCode" id="cb16"><html><body><pre class="r"><span class="no">pbmc_small_polished</span> <span class="kw">%&gt;%</span>
  <span class="fu"><a href="../reference/join_transcripts.html">join_transcripts</a></span>(<span class="kw">transcripts</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"HLA-DRA"</span>, <span class="st">"LYZ"</span>)) <span class="kw">%&gt;%</span>
  <span class="fu"><a href="../reference/ggplot2-methods.html">ggplot</a></span>(<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="no">groups</span>, <span class="no">abundance_RNA</span> + <span class="fl">1</span>, <span class="kw">fill</span> <span class="kw">=</span> <span class="no">groups</span>)) +
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html">geom_boxplot</a></span>(<span class="kw">outlier.shape</span> <span class="kw">=</span> <span class="fl">NA</span>) +
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_jitter.html">geom_jitter</a></span>(<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="kw">size</span> <span class="kw">=</span> <span class="no">nCount_RNA</span>), <span class="kw">alpha</span> <span class="kw">=</span> <span class="fl">0.5</span>, <span class="kw">width</span> <span class="kw">=</span> <span class="fl">0.2</span>) +
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_y_log10</a></span>() +
  <span class="no">my_theme</span></pre></body></html></div>
<p><img src="introduction_files/figure-html/unnamed-chunk-14-1.png" width="700"></p>
</div>
<div id="preprocess-the-dataset" class="section level1">
<h1 class="hasAnchor">
<a href="#preprocess-the-dataset" class="anchor"></a>Preprocess the dataset</h1>
<p>Also you can treat the object as Seurat object and proceed with data processing.</p>
<div class="sourceCode" id="cb17"><html><body><pre class="r"><span class="no">pbmc_small_pca</span> <span class="kw">&lt;-</span>
  <span class="no">pbmc_small_polished</span> <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/SCTransform.html">SCTransform</a></span>(<span class="kw">verbose</span> <span class="kw">=</span> <span class="fl">FALSE</span>) <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/FindVariableFeatures.html">FindVariableFeatures</a></span>(<span class="kw">verbose</span> <span class="kw">=</span> <span class="fl">FALSE</span>) <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/RunPCA.html">RunPCA</a></span>(<span class="kw">verbose</span> <span class="kw">=</span> <span class="fl">FALSE</span>)

<span class="no">pbmc_small_pca</span></pre></body></html></div>
<pre><code>## # A tibble abstraction: 80 x 19
##    cell  orig.ident nCount_RNA nFeature_RNA RNA_snn_res.0.8 letter.idents groups
##    &lt;chr&gt; &lt;fct&gt;           &lt;dbl&gt;        &lt;int&gt; &lt;fct&gt;           &lt;fct&gt;         &lt;chr&gt; 
##  1 ATGC… SeuratPro…         70           47 0               A             g2    
##  2 CATG… SeuratPro…         85           52 0               A             g1    
##  3 GAAC… SeuratPro…         87           50 1               B             g2    
##  4 TGAC… SeuratPro…        127           56 0               A             g2    
##  5 AGTC… SeuratPro…        173           53 0               A             g2    
##  6 TCTG… SeuratPro…         70           48 0               A             g1    
##  7 TGGT… SeuratPro…         64           36 0               A             g1    
##  8 GCAG… SeuratPro…         72           45 0               A             g1    
##  9 GATA… SeuratPro…         52           36 0               A             g1    
## 10 AATG… SeuratPro…        100           41 0               A             g1    
## # … with 70 more rows, and 12 more variables: RNA_snn_res.1 &lt;fct&gt;, file &lt;chr&gt;,
## #   sample &lt;chr&gt;, nCount_SCT &lt;dbl&gt;, nFeature_SCT &lt;int&gt;, PC_1 &lt;dbl&gt;, PC_2 &lt;dbl&gt;,
## #   PC_3 &lt;dbl&gt;, PC_4 &lt;dbl&gt;, PC_5 &lt;dbl&gt;, tSNE_1 &lt;dbl&gt;, tSNE_2 &lt;dbl&gt;</code></pre>
<p>If a tool is not included in the tidyseurat collection, we can use <code>as_tibble</code> to permanently convert <code>tidyseurat</code> into tibble.</p>
<div class="sourceCode" id="cb19"><html><body><pre class="r"><span class="no">pbmc_small_pca</span> <span class="kw">%&gt;%</span>
  <span class="fu"><a href="../reference/tibble-methods.html">as_tibble</a></span>() <span class="kw">%&gt;%</span>
  <span class="fu"><a href="../reference/dplyr-methods.html">select</a></span>(<span class="fu"><a href="https://dplyr.tidyverse.org/reference/reexports.html">contains</a></span>(<span class="st">"PC"</span>), <span class="fu"><a href="https://dplyr.tidyverse.org/reference/reexports.html">everything</a></span>()) <span class="kw">%&gt;%</span>
  <span class="kw pkg">GGally</span><span class="kw ns">::</span><span class="fu"><a href="https://ggobi.github.io/ggally/reference/ggpairs.html">ggpairs</a></span>(<span class="kw">columns</span> <span class="kw">=</span> <span class="fl">1</span>:<span class="fl">5</span>, <span class="kw pkg">ggplot2</span><span class="kw ns">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="kw">colour</span> <span class="kw">=</span> <span class="no">groups</span>)) +
  <span class="no">my_theme</span></pre></body></html></div>
<p><img src="introduction_files/figure-html/pc_plot-1.png" width="700"></p>
</div>
<div id="identify-clusters" class="section level1">
<h1 class="hasAnchor">
<a href="#identify-clusters" class="anchor"></a>Identify clusters</h1>
<p>We proceed with cluster identification with Seurat.</p>
<div class="sourceCode" id="cb20"><html><body><pre class="r"><span class="no">pbmc_small_cluster</span> <span class="kw">&lt;-</span>
  <span class="no">pbmc_small_pca</span> <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/FindNeighbors.html">FindNeighbors</a></span>(<span class="kw">verbose</span> <span class="kw">=</span> <span class="fl">FALSE</span>) <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/FindClusters.html">FindClusters</a></span>(<span class="kw">method</span> <span class="kw">=</span> <span class="st">"igraph"</span>, <span class="kw">verbose</span> <span class="kw">=</span> <span class="fl">FALSE</span>)

<span class="no">pbmc_small_cluster</span></pre></body></html></div>
<pre><code>## # A tibble abstraction: 80 x 21
##    cell  orig.ident nCount_RNA nFeature_RNA RNA_snn_res.0.8 letter.idents groups
##    &lt;chr&gt; &lt;fct&gt;           &lt;dbl&gt;        &lt;int&gt; &lt;fct&gt;           &lt;fct&gt;         &lt;chr&gt; 
##  1 ATGC… SeuratPro…         70           47 0               A             g2    
##  2 CATG… SeuratPro…         85           52 0               A             g1    
##  3 GAAC… SeuratPro…         87           50 1               B             g2    
##  4 TGAC… SeuratPro…        127           56 0               A             g2    
##  5 AGTC… SeuratPro…        173           53 0               A             g2    
##  6 TCTG… SeuratPro…         70           48 0               A             g1    
##  7 TGGT… SeuratPro…         64           36 0               A             g1    
##  8 GCAG… SeuratPro…         72           45 0               A             g1    
##  9 GATA… SeuratPro…         52           36 0               A             g1    
## 10 AATG… SeuratPro…        100           41 0               A             g1    
## # … with 70 more rows, and 14 more variables: RNA_snn_res.1 &lt;fct&gt;, file &lt;chr&gt;,
## #   sample &lt;chr&gt;, nCount_SCT &lt;dbl&gt;, nFeature_SCT &lt;int&gt;, SCT_snn_res.0.8 &lt;fct&gt;,
## #   seurat_clusters &lt;fct&gt;, PC_1 &lt;dbl&gt;, PC_2 &lt;dbl&gt;, PC_3 &lt;dbl&gt;, PC_4 &lt;dbl&gt;,
## #   PC_5 &lt;dbl&gt;, tSNE_1 &lt;dbl&gt;, tSNE_2 &lt;dbl&gt;</code></pre>
<p>Now we can interrogate the object as if it was a regular tibble data frame.</p>
<div class="sourceCode" id="cb22"><html><body><pre class="r"><span class="no">pbmc_small_cluster</span> <span class="kw">%&gt;%</span>
  <span class="kw pkg">tidyseurat</span><span class="kw ns">::</span><span class="fu"><a href="../reference/count.html">count</a></span>(<span class="no">groups</span>, <span class="no">seurat_clusters</span>)</pre></body></html></div>
<pre><code>## # A tibble: 8 x 3
##   groups seurat_clusters     n
##   &lt;chr&gt;  &lt;fct&gt;           &lt;int&gt;
## 1 g1     0                  17
## 2 g1     1                  14
## 3 g1     2                   9
## 4 g1     3                   4
## 5 g2     0                  13
## 6 g2     1                  12
## 7 g2     2                   6
## 8 g2     3                   5</code></pre>
<p>We can identify cluster markers using Seurat.</p>
<div class="sourceCode" id="cb24"><html><body><pre class="r"><span class="co"># Identify top 10 markers per cluster</span>
<span class="no">markers</span> <span class="kw">&lt;-</span>
  <span class="no">pbmc_small_cluster</span> <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/FindAllMarkers.html">FindAllMarkers</a></span>(<span class="kw">only.pos</span> <span class="kw">=</span> <span class="fl">TRUE</span>, <span class="kw">min.pct</span> <span class="kw">=</span> <span class="fl">0.25</span>, <span class="kw">thresh.use</span> <span class="kw">=</span> <span class="fl">0.25</span>) <span class="kw">%&gt;%</span>
  <span class="fu"><a href="../reference/dplyr-methods.html">group_by</a></span>(<span class="no">cluster</span>) <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/top_n.html">top_n</a></span>(<span class="fl">10</span>, <span class="no">avg_logFC</span>)

<span class="co"># Plot heatmap</span>
<span class="no">pbmc_small_cluster</span> <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/DoHeatmap.html">DoHeatmap</a></span>(
    <span class="kw">features</span> <span class="kw">=</span> <span class="no">markers</span>$<span class="no">gene</span>,
    <span class="kw">group.colors</span> <span class="kw">=</span> <span class="no">friendly_cols</span>
  )</pre></body></html></div>
<p><img src="introduction_files/figure-html/unnamed-chunk-15-1.png" width="700"></p>
</div>
<div id="reduce-dimensions" class="section level1">
<h1 class="hasAnchor">
<a href="#reduce-dimensions" class="anchor"></a>Reduce dimensions</h1>
<p>We can calculate the first 3 UMAP dimensions using the Seurat framework.</p>
<div class="sourceCode" id="cb25"><html><body><pre class="r"><span class="no">pbmc_small_UMAP</span> <span class="kw">&lt;-</span>
  <span class="no">pbmc_small_cluster</span> <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/RunUMAP.html">RunUMAP</a></span>(<span class="kw">reduction</span> <span class="kw">=</span> <span class="st">"pca"</span>, <span class="kw">dims</span> <span class="kw">=</span> <span class="fl">1</span>:<span class="fl">15</span>, <span class="kw">n.components</span> <span class="kw">=</span> <span class="fl">3L</span>, )</pre></body></html></div>
<p>And we can plot them using 3D plot using plotly.</p>
<div class="sourceCode" id="cb26"><html><body><pre class="r"><span class="no">pbmc_small_UMAP</span> <span class="kw">%&gt;%</span>
  <span class="fu"><a href="../reference/plot_ly.html">plot_ly</a></span>(
    <span class="kw">x</span> <span class="kw">=</span> ~<span class="no">`UMAP_1`</span>,
    <span class="kw">y</span> <span class="kw">=</span> ~<span class="no">`UMAP_2`</span>,
    <span class="kw">z</span> <span class="kw">=</span> ~<span class="no">`UMAP_3`</span>,
    <span class="kw">color</span> <span class="kw">=</span> ~<span class="no">seurat_clusters</span>,
    <span class="kw">colors</span> <span class="kw">=</span> <span class="no">friendly_cols</span>[<span class="fl">1</span>:<span class="fl">4</span>]
  )</pre></body></html></div>
<div class="figure">
<img src="../man/figures/plotly.png" alt=""><p class="caption">screenshot plotly</p>
</div>
<div id="cell-type-prediction" class="section level2">
<h2 class="hasAnchor">
<a href="#cell-type-prediction" class="anchor"></a>Cell type prediction</h2>
<p>We can infer cell type identities using <em>SingleR</em> <span class="citation">(Aran et al. 2019)</span> and manipulate the output using tidyverse.</p>
<div class="sourceCode" id="cb27"><html><body><pre class="r"><span class="co"># Get cell type reference data</span>
<span class="no">blueprint</span> <span class="kw">&lt;-</span> <span class="kw pkg">celldex</span><span class="kw ns">::</span><span class="fu">BlueprintEncodeData</span>()

<span class="co"># Infer cell identities</span>
<span class="no">cell_type_df</span> <span class="kw">&lt;-</span>
  <span class="no">pbmc_small_UMAP</span>@<span class="kw">assays</span><span class="kw">[[</span><span class="st">"SCT"</span>]]@<span class="kw">counts</span> <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/r/base/Log.html">log1p</a></span>() <span class="kw">%&gt;%</span>
  <span class="kw pkg">Matrix</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/Matrix.html">Matrix</a></span>(<span class="kw">sparse</span> <span class="kw">=</span> <span class="fl">TRUE</span>) <span class="kw">%&gt;%</span>
  <span class="kw pkg">SingleR</span><span class="kw ns">::</span><span class="fu">SingleR</span>(
    <span class="kw">ref</span> <span class="kw">=</span> <span class="no">blueprint</span>,
    <span class="kw">labels</span> <span class="kw">=</span> <span class="no">blueprint</span>$<span class="no">label.main</span>,
    <span class="kw">method</span> <span class="kw">=</span> <span class="st">"single"</span>
  ) <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span>() <span class="kw">%&gt;%</span>
  <span class="fu"><a href="../reference/tibble-methods.html">as_tibble</a></span>(<span class="kw">rownames</span> <span class="kw">=</span> <span class="st">"cell"</span>) <span class="kw">%&gt;%</span>
  <span class="fu"><a href="../reference/dplyr-methods.html">select</a></span>(<span class="no">cell</span>, <span class="no">first.labels</span>)</pre></body></html></div>
<div class="sourceCode" id="cb28"><html><body><pre class="r"><span class="co"># Join UMAP and cell type info</span>
<span class="no">pbmc_small_cell_type</span> <span class="kw">&lt;-</span>
  <span class="no">pbmc_small_UMAP</span> <span class="kw">%&gt;%</span>
  <span class="fu"><a href="../reference/dplyr-methods.html">left_join</a></span>(<span class="no">cell_type_df</span>, <span class="kw">by</span> <span class="kw">=</span> <span class="st">"cell"</span>)

<span class="co"># Reorder columns</span>
<span class="no">pbmc_small_cell_type</span> <span class="kw">%&gt;%</span>
  <span class="kw pkg">tidyseurat</span><span class="kw ns">::</span><span class="fu"><a href="../reference/dplyr-methods.html">select</a></span>(<span class="no">cell</span>, <span class="no">first.labels</span>, <span class="fu"><a href="https://dplyr.tidyverse.org/reference/reexports.html">everything</a></span>())</pre></body></html></div>
<pre><code>## # A tibble abstraction: 80 x 25
##    cell  first.labels orig.ident nCount_RNA nFeature_RNA RNA_snn_res.0.8
##    &lt;chr&gt; &lt;chr&gt;        &lt;fct&gt;           &lt;dbl&gt;        &lt;int&gt; &lt;fct&gt;          
##  1 ATGC… CD4+ T-cells SeuratPro…         70           47 0              
##  2 CATG… CD8+ T-cells SeuratPro…         85           52 0              
##  3 GAAC… CD8+ T-cells SeuratPro…         87           50 1              
##  4 TGAC… CD4+ T-cells SeuratPro…        127           56 0              
##  5 AGTC… CD4+ T-cells SeuratPro…        173           53 0              
##  6 TCTG… CD4+ T-cells SeuratPro…         70           48 0              
##  7 TGGT… CD4+ T-cells SeuratPro…         64           36 0              
##  8 GCAG… CD4+ T-cells SeuratPro…         72           45 0              
##  9 GATA… CD4+ T-cells SeuratPro…         52           36 0              
## 10 AATG… CD4+ T-cells SeuratPro…        100           41 0              
## # … with 70 more rows, and 19 more variables: letter.idents &lt;fct&gt;,
## #   groups &lt;chr&gt;, RNA_snn_res.1 &lt;fct&gt;, file &lt;chr&gt;, sample &lt;chr&gt;,
## #   nCount_SCT &lt;dbl&gt;, nFeature_SCT &lt;int&gt;, SCT_snn_res.0.8 &lt;fct&gt;,
## #   seurat_clusters &lt;fct&gt;, PC_1 &lt;dbl&gt;, PC_2 &lt;dbl&gt;, PC_3 &lt;dbl&gt;, PC_4 &lt;dbl&gt;,
## #   PC_5 &lt;dbl&gt;, tSNE_1 &lt;dbl&gt;, tSNE_2 &lt;dbl&gt;, UMAP_1 &lt;dbl&gt;, UMAP_2 &lt;dbl&gt;,
## #   UMAP_3 &lt;dbl&gt;</code></pre>
<p>We can easily summarise the results. For example, we can see how cell type classification overlaps with cluster classification.</p>
<div class="sourceCode" id="cb30"><html><body><pre class="r"><span class="no">pbmc_small_cell_type</span> <span class="kw">%&gt;%</span>
  <span class="fu"><a href="../reference/count.html">count</a></span>(<span class="no">seurat_clusters</span>, <span class="no">first.labels</span>)</pre></body></html></div>
<pre><code>## # A tibble: 9 x 3
##   seurat_clusters first.labels     n
##   &lt;fct&gt;           &lt;chr&gt;        &lt;int&gt;
## 1 0               CD4+ T-cells     8
## 2 0               CD8+ T-cells    10
## 3 0               NK cells        12
## 4 1               Macrophages      1
## 5 1               Monocytes       25
## 6 2               B-cells         10
## 7 2               Macrophages      1
## 8 2               Monocytes        4
## 9 3               Erythrocytes     9</code></pre>
<p>We can easily reshape the data for building information-rich faceted plots.</p>
<div class="sourceCode" id="cb32"><html><body><pre class="r"><span class="no">pbmc_small_cell_type</span> <span class="kw">%&gt;%</span>

  <span class="co"># Reshape and add classifier column</span>
  <span class="fu"><a href="../reference/pivot_longer.html">pivot_longer</a></span>(
    <span class="kw">cols</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="no">seurat_clusters</span>, <span class="no">first.labels</span>),
    <span class="kw">names_to</span> <span class="kw">=</span> <span class="st">"classifier"</span>, <span class="kw">values_to</span> <span class="kw">=</span> <span class="st">"label"</span>
  ) <span class="kw">%&gt;%</span>

  <span class="co"># UMAP plots for cell type and cluster</span>
  <span class="fu"><a href="../reference/ggplot2-methods.html">ggplot</a></span>(<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="no">UMAP_1</span>, <span class="no">UMAP_2</span>, <span class="kw">color</span> <span class="kw">=</span> <span class="no">label</span>)) +
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span>() +
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html">facet_wrap</a></span>(~<span class="no">classifier</span>) +
  <span class="no">my_theme</span></pre></body></html></div>
<p><img src="introduction_files/figure-html/unnamed-chunk-19-1.png" width="700"></p>
<p>We can easily plot gene correlation per cell category, adding multi-layer annotations.</p>
<div class="sourceCode" id="cb33"><html><body><pre class="r"><span class="no">pbmc_small_cell_type</span> <span class="kw">%&gt;%</span>

  <span class="co"># Add some mitochondrial abundance values</span>
  <span class="fu"><a href="../reference/dplyr-methods.html">mutate</a></span>(<span class="kw">mitochondrial</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span>(<span class="fu"><a href="https://dplyr.tidyverse.org/reference/context.html">n</a></span>())) <span class="kw">%&gt;%</span>

  <span class="co"># Plot correlation</span>
  <span class="fu"><a href="../reference/join_transcripts.html">join_transcripts</a></span>(<span class="kw">transcripts</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"CST3"</span>, <span class="st">"LYZ"</span>), <span class="kw">shape</span> <span class="kw">=</span> <span class="st">"wide"</span>) <span class="kw">%&gt;%</span>
  <span class="fu"><a href="../reference/ggplot2-methods.html">ggplot</a></span>(<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="no">CST3</span> + <span class="fl">1</span>, <span class="no">LYZ</span> + <span class="fl">1</span>, <span class="kw">color</span> <span class="kw">=</span> <span class="no">groups</span>, <span class="kw">size</span> <span class="kw">=</span> <span class="no">mitochondrial</span>)) +
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span>() +
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html">facet_wrap</a></span>(~<span class="no">first.labels</span>, <span class="kw">scales</span> <span class="kw">=</span> <span class="st">"free"</span>) +
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_x_log10</a></span>() +
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_y_log10</a></span>() +
  <span class="no">my_theme</span></pre></body></html></div>
<p><img src="introduction_files/figure-html/unnamed-chunk-20-1.png" width="700"></p>
</div>
</div>
<div id="nested-analyses" class="section level1">
<h1 class="hasAnchor">
<a href="#nested-analyses" class="anchor"></a>Nested analyses</h1>
<p>A powerful tool we can use with tidyseurat is <code>nest</code>. We can easily perform independent analyses on subsets of the dataset. First we classify cell types in lymphoid and myeloid; then, nest based on the new classification</p>
<div class="sourceCode" id="cb34"><html><body><pre class="r"><span class="no">pbmc_small_nested</span> <span class="kw">&lt;-</span>
  <span class="no">pbmc_small_cell_type</span> <span class="kw">%&gt;%</span>
  <span class="fu"><a href="../reference/dplyr-methods.html">filter</a></span>(<span class="no">first.labels</span> <span class="kw">!=</span> <span class="st">"Erythrocytes"</span>) <span class="kw">%&gt;%</span>
  <span class="fu"><a href="../reference/dplyr-methods.html">mutate</a></span>(<span class="kw">cell_class</span> <span class="kw">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/if_else.html">if_else</a></span>(<span class="no">`first.labels`</span> <span class="kw">%in%</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"Macrophages"</span>, <span class="st">"Monocytes"</span>), <span class="st">"myeloid"</span>, <span class="st">"lymphoid"</span>)) <span class="kw">%&gt;%</span>
  <span class="fu"><a href="../reference/tidyr-methods.html">nest</a></span>(<span class="kw">data</span> <span class="kw">=</span> -<span class="no">cell_class</span>)

<span class="no">pbmc_small_nested</span></pre></body></html></div>
<pre><code>## # A tibble: 2 x 2
##   cell_class data      
##   &lt;chr&gt;      &lt;list&gt;    
## 1 lymphoid   &lt;tidysert&gt;
## 2 myeloid    &lt;tidysert&gt;</code></pre>
<p>Now we can independently for the lymphoid and myeloid subsets (i) find variable features, (ii) reduce dimensions, and (iii) cluster using both tidyverse and SingleCellExperiment seamlessly.</p>
<div class="sourceCode" id="cb36"><html><body><pre class="r"><span class="no">pbmc_small_nested_reanalysed</span> <span class="kw">&lt;-</span>
  <span class="no">pbmc_small_nested</span> <span class="kw">%&gt;%</span>
  <span class="fu"><a href="../reference/dplyr-methods.html">mutate</a></span>(<span class="kw">data</span> <span class="kw">=</span> <span class="fu"><a href="https://purrr.tidyverse.org/reference/map.html">map</a></span>(
    <span class="no">data</span>, ~ <span class="no">.x</span> <span class="kw">%&gt;%</span>
      <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/FindVariableFeatures.html">FindVariableFeatures</a></span>(<span class="kw">verbose</span> <span class="kw">=</span> <span class="fl">FALSE</span>) <span class="kw">%&gt;%</span>
      <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/RunPCA.html">RunPCA</a></span>(<span class="kw">npcs</span> <span class="kw">=</span> <span class="fl">10</span>, <span class="kw">verbose</span> <span class="kw">=</span> <span class="fl">FALSE</span>) <span class="kw">%&gt;%</span>
      <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/FindNeighbors.html">FindNeighbors</a></span>(<span class="kw">verbose</span> <span class="kw">=</span> <span class="fl">FALSE</span>) <span class="kw">%&gt;%</span>
      <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/FindClusters.html">FindClusters</a></span>(<span class="kw">method</span> <span class="kw">=</span> <span class="st">"igraph"</span>, <span class="kw">verbose</span> <span class="kw">=</span> <span class="fl">FALSE</span>) <span class="kw">%&gt;%</span>
      <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/RunUMAP.html">RunUMAP</a></span>(<span class="kw">reduction</span> <span class="kw">=</span> <span class="st">"pca"</span>, <span class="kw">dims</span> <span class="kw">=</span> <span class="fl">1</span>:<span class="fl">10</span>, <span class="kw">n.components</span> <span class="kw">=</span> <span class="fl">3L</span>, <span class="kw">verbose</span> <span class="kw">=</span> <span class="fl">FALSE</span>)
  ))

<span class="no">pbmc_small_nested_reanalysed</span></pre></body></html></div>
<pre><code>## # A tibble: 2 x 2
##   cell_class data      
##   &lt;chr&gt;      &lt;list&gt;    
## 1 lymphoid   &lt;tidysert&gt;
## 2 myeloid    &lt;tidysert&gt;</code></pre>
<p>Now we can unnest and plot the new classification.</p>
<div class="sourceCode" id="cb38"><html><body><pre class="r"><span class="no">pbmc_small_nested_reanalysed</span> <span class="kw">%&gt;%</span>

  <span class="co"># Convert to tibble otherwise Seurat drops reduced dimensions when unifying data sets.</span>
  <span class="fu"><a href="../reference/dplyr-methods.html">mutate</a></span>(<span class="kw">data</span> <span class="kw">=</span> <span class="fu"><a href="https://purrr.tidyverse.org/reference/map.html">map</a></span>(<span class="no">data</span>, ~ <span class="no">.x</span> <span class="kw">%&gt;%</span> <span class="fu"><a href="../reference/tibble-methods.html">as_tibble</a></span>())) <span class="kw">%&gt;%</span>
  <span class="fu"><a href="../reference/tidyr-methods.html">unnest</a></span>(<span class="no">data</span>) <span class="kw">%&gt;%</span>

  <span class="co"># Define unique clusters</span>
  <span class="fu"><a href="../reference/tidyr-methods.html">unite</a></span>(<span class="st">"cluster"</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="no">cell_class</span>, <span class="no">seurat_clusters</span>), <span class="kw">remove</span> <span class="kw">=</span> <span class="fl">FALSE</span>) <span class="kw">%&gt;%</span>

  <span class="co"># Plotting</span>
  <span class="fu"><a href="../reference/ggplot2-methods.html">ggplot</a></span>(<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="no">UMAP_1</span>, <span class="no">UMAP_2</span>, <span class="kw">color</span> <span class="kw">=</span> <span class="no">cluster</span>)) +
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span>() +
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html">facet_wrap</a></span>(~<span class="no">cell_class</span>) +
  <span class="no">my_theme</span></pre></body></html></div>
<p><img src="introduction_files/figure-html/unnamed-chunk-23-1.png" width="700"></p>
<p>We can perform a large number of functional analyses on data subsets. For example, we can identify intra-sample cell-cell interactions using <em>SingleCellSignalR</em> <span class="citation">(Cabello-Aguilar et al. 2020)</span>, and then compare whether interactions are stronger or weaker across conditions. The code below demonstrates how this analysis could be performed. It won’t work with this small example dataset as we have just two samples (one for each condition). But some example output is shown below and you can imagine how you can use tidyverse on the output to perform t-tests and visualisation.</p>
<div class="sourceCode" id="cb39"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">SingleCellSignalR</span>)

<span class="no">pbmc_small_nested_interactions</span> <span class="kw">&lt;-</span>
  <span class="no">pbmc_small_nested_reanalysed</span> <span class="kw">%&gt;%</span>

  <span class="co"># Unnest based on cell category</span>
  <span class="fu"><a href="../reference/tidyr-methods.html">unnest</a></span>(<span class="no">data</span>) <span class="kw">%&gt;%</span>

  <span class="co"># Create unambiguous clusters</span>
  <span class="fu"><a href="../reference/dplyr-methods.html">mutate</a></span>(<span class="kw">integrated_clusters</span> <span class="kw">=</span> <span class="no">first.labels</span> <span class="kw">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">as.factor</a></span>() <span class="kw">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/base/integer.html">as.integer</a></span>()) <span class="kw">%&gt;%</span>

  <span class="co"># Nest based on sample</span>
  <span class="kw pkg">tidyseurat</span><span class="kw ns">::</span><span class="fu"><a href="../reference/tidyr-methods.html">nest</a></span>(<span class="kw">data</span> <span class="kw">=</span> -<span class="no">sample</span>) <span class="kw">%&gt;%</span>
  <span class="kw pkg">tidyseurat</span><span class="kw ns">::</span><span class="fu"><a href="../reference/dplyr-methods.html">mutate</a></span>(<span class="kw">interactions</span> <span class="kw">=</span> <span class="fu"><a href="https://purrr.tidyverse.org/reference/map.html">map</a></span>(<span class="no">data</span>, ~ {

    <span class="co"># Produce variables. Yuck!</span>
    <span class="no">cluster</span> <span class="kw">&lt;-</span> <span class="no">.x</span>@<span class="kw">meta.data</span>$<span class="no">integrated_clusters</span>
    <span class="no">data</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span>(<span class="no">.x</span><span class="kw">[[</span><span class="st">"SCT"</span>]]@<span class="kw">data</span>)

    <span class="co"># Ligand/Receptor analysis using SingleCellSignalR</span>
    <span class="no">data</span> <span class="kw">%&gt;%</span>
      <span class="fu"><a href="https://rdrr.io/pkg/SingleCellSignalR/man/cell_signaling.html">cell_signaling</a></span>(<span class="kw">genes</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(<span class="no">data</span>), <span class="kw">cluster</span> <span class="kw">=</span> <span class="no">cluster</span>) <span class="kw">%&gt;%</span>
      <span class="fu"><a href="https://rdrr.io/pkg/SingleCellSignalR/man/inter_network.html">inter_network</a></span>(<span class="kw">data</span> <span class="kw">=</span> <span class="no">data</span>, <span class="kw">signal</span> <span class="kw">=</span> <span class="no">.</span>, <span class="kw">genes</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(<span class="no">data</span>), <span class="kw">cluster</span> <span class="kw">=</span> <span class="no">cluster</span>) <span class="kw">%$%</span>
      <span class="no">`individual-networks`</span> <span class="kw">%&gt;%</span>
      <span class="fu"><a href="https://purrr.tidyverse.org/reference/map.html">map_dfr</a></span>(~ <span class="fu"><a href="../reference/dplyr-methods.html">bind_rows</a></span>(<span class="fu"><a href="../reference/tibble-methods.html">as_tibble</a></span>(<span class="no">.x</span>)))
  }))

<span class="no">pbmc_small_nested_interactions</span> <span class="kw">%&gt;%</span>
  <span class="fu"><a href="../reference/dplyr-methods.html">select</a></span>(-<span class="no">data</span>) <span class="kw">%&gt;%</span>
  <span class="fu"><a href="../reference/tidyr-methods.html">unnest</a></span>(<span class="no">interactions</span>)</pre></body></html></div>
<p>If the data set was not so small, and interactions could be identified, you would see something as below.</p>
<pre><code>## # A tibble: 100 x 9
##    sample ligand receptor ligand.name receptor.name origin destination
##    &lt;chr&gt;  &lt;chr&gt;  &lt;chr&gt;    &lt;chr&gt;       &lt;chr&gt;         &lt;chr&gt;  &lt;chr&gt;      
##  1 sampl… clust… cluster… PTMA        VIPR1         clust… cluster 2  
##  2 sampl… clust… cluster… B2M         KLRD1         clust… cluster 2  
##  3 sampl… clust… cluster… IL16        CD4           clust… cluster 2  
##  4 sampl… clust… cluster… HLA-B       KLRD1         clust… cluster 2  
##  5 sampl… clust… cluster… CALM1       VIPR1         clust… cluster 2  
##  6 sampl… clust… cluster… HLA-E       KLRD1         clust… cluster 2  
##  7 sampl… clust… cluster… GNAS        VIPR1         clust… cluster 2  
##  8 sampl… clust… cluster… B2M         HFE           clust… cluster 2  
##  9 sampl… clust… cluster… PTMA        VIPR1         clust… cluster 3  
## 10 sampl… clust… cluster… CALM1       VIPR1         clust… cluster 3  
## # … with 90 more rows, and 2 more variables: interaction.type &lt;chr&gt;,
## #   LRscore &lt;dbl&gt;</code></pre>
</div>
<div id="session-info" class="section level1">
<h1 class="hasAnchor">
<a href="#session-info" class="anchor"></a>Session Info</h1>
<div class="sourceCode" id="cb41"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span>()</pre></body></html></div>
<pre><code>## R version 4.0.0 (2020-04-24)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Ubuntu 18.04.5 LTS
## 
## Matrix products: default
## BLAS/LAPACK: /usr/lib/x86_64-linux-gnu/libopenblasp-r0.2.20.so
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=C             
##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
## [1] stats     graphics  grDevices utils     datasets  methods   base     
## 
## other attached packages:
## [1] tidyseurat_0.1.11 Seurat_3.1.5      ggplot2_3.3.1     magrittr_1.5     
## [5] purrr_0.3.4       tidyr_1.1.0       dplyr_1.0.0       knitr_1.28       
## 
## loaded via a namespace (and not attached):
##   [1] Rtsne_0.15                  colorspace_1.4-1           
##   [3] ellipsis_0.3.1              ggridges_0.5.2             
##   [5] rprojroot_1.3-2             XVector_0.28.0             
##   [7] GenomicRanges_1.40.0        fs_1.4.1                   
##   [9] farver_2.0.3                leiden_0.3.3               
##  [11] listenv_0.8.0               ggrepel_0.8.2              
##  [13] RSpectra_0.16-0             fansi_0.4.1                
##  [15] codetools_0.2-16            splines_4.0.0              
##  [17] jsonlite_1.6.1              ica_1.0-2                  
##  [19] cluster_2.1.0               png_0.1-7                  
##  [21] pheatmap_1.0.12             uwot_0.1.8                 
##  [23] sctransform_0.2.1           compiler_4.0.0             
##  [25] httr_1.4.1                  backports_1.1.7            
##  [27] assertthat_0.2.1            Matrix_1.2-18              
##  [29] lazyeval_0.2.2              limma_3.44.3               
##  [31] cli_2.0.2                   htmltools_0.4.0            
##  [33] tools_4.0.0                 rsvd_1.0.3                 
##  [35] igraph_1.2.5                gtable_0.3.0               
##  [37] glue_1.4.1                  GenomeInfoDbData_1.2.3     
##  [39] RANN_2.6.1                  reshape2_1.4.4             
##  [41] rappdirs_0.3.1              Rcpp_1.0.4.6               
##  [43] Biobase_2.48.0              pkgdown_1.5.1              
##  [45] vctrs_0.3.1                 ape_5.4                    
##  [47] nlme_3.1-148                lmtest_0.9-37              
##  [49] xfun_0.14                   stringr_1.4.0              
##  [51] globals_0.12.5              lifecycle_0.2.0            
##  [53] irlba_2.3.3                 future_1.17.0              
##  [55] edgeR_3.30.3                MASS_7.3-51.6              
##  [57] zlibbioc_1.34.0             zoo_1.8-8                  
##  [59] scales_1.1.1                parallel_4.0.0             
##  [61] SummarizedExperiment_1.18.2 RColorBrewer_1.1-2         
##  [63] SingleCellExperiment_1.10.1 yaml_2.2.1                 
##  [65] memoise_1.1.0               reticulate_1.16            
##  [67] pbapply_1.4-2               gridExtra_2.3              
##  [69] reshape_0.8.8               stringi_1.4.6              
##  [71] S4Vectors_0.26.1            desc_1.2.0                 
##  [73] BiocGenerics_0.34.0         GenomeInfoDb_1.24.2        
##  [75] rlang_0.4.6                 pkgconfig_2.0.3            
##  [77] bitops_1.0-6                matrixStats_0.56.0         
##  [79] evaluate_0.14               lattice_0.20-41            
##  [81] ROCR_1.0-11                 patchwork_1.0.0            
##  [83] htmlwidgets_1.5.1           labeling_0.3               
##  [85] cowplot_1.0.0               tidyselect_1.1.0           
##  [87] GGally_2.0.0                RcppAnnoy_0.0.16           
##  [89] plyr_1.8.6                  R6_2.4.1                   
##  [91] IRanges_2.22.2              generics_0.0.2             
##  [93] DelayedArray_0.14.1         pillar_1.4.4               
##  [95] withr_2.2.0                 fitdistrplus_1.1-1         
##  [97] survival_3.1-12             RCurl_1.98-1.2             
##  [99] tibble_3.0.1                future.apply_1.5.0         
## [101] tsne_0.1-3                  crayon_1.3.4               
## [103] KernSmooth_2.23-17          utf8_1.1.4                 
## [105] dittoSeq_1.0.2              plotly_4.9.2.1             
## [107] rmarkdown_2.2               locfit_1.5-9.4             
## [109] grid_4.0.0                  data.table_1.12.8          
## [111] digest_0.6.25               stats4_4.0.0               
## [113] munsell_0.5.0               viridisLite_0.3.0</code></pre>
</div>
<div id="references" class="section level1 unnumbered">
<h1 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h1>
<div id="refs" class="references">
<div id="ref-aran2019reference">
<p>Aran, Dvir, Agnieszka P Looney, Leqian Liu, Esther Wu, Valerie Fong, Austin Hsu, Suzanna Chak, et al. 2019. “Reference-Based Analysis of Lung Single-Cell Sequencing Reveals a Transitional Profibrotic Macrophage.” <em>Nature Immunology</em> 20 (2): 163–72.</p>
</div>
<div id="ref-butler2018integrating">
<p>Butler, Andrew, Paul Hoffman, Peter Smibert, Efthymia Papalexi, and Rahul Satija. 2018. “Integrating Single-Cell Transcriptomic Data Across Different Conditions, Technologies, and Species.” <em>Nature Biotechnology</em> 36 (5): 411–20.</p>
</div>
<div id="ref-cabello2020singlecellsignalr">
<p>Cabello-Aguilar, Simon, Mélissa Alame, Fabien Kon-Sun-Tack, Caroline Fau, Matthieu Lacroix, and Jacques Colinge. 2020. “SingleCellSignalR: Inference of Intercellular Networks from Single-Cell Transcriptomics.” <em>Nucleic Acids Research</em> 48 (10): e55–e55.</p>
</div>
<div id="ref-stuart2019comprehensive">
<p>Stuart, Tim, Andrew Butler, Paul Hoffman, Christoph Hafemeister, Efthymia Papalexi, William M Mauck III, Yuhan Hao, Marlon Stoeckius, Peter Smibert, and Rahul Satija. 2019. “Comprehensive Integration of Single-Cell Data.” <em>Cell</em> 177 (7): 1888–1902.</p>
</div>
<div id="ref-wickham2019welcome">
<p>Wickham, Hadley, Mara Averick, Jennifer Bryan, Winston Chang, Lucy D’Agostino McGowan, Romain François, Garrett Grolemund, et al. 2019. “Welcome to the Tidyverse.” <em>Journal of Open Source Software</em> 4 (43): 1686.</p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Stefano Mangiola.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
